File: Unity/Assets/ML/resnet18/Runtime/ResNet18HubPredictor.cs
/* 
*   resnet18
*   Copyright (c) 2021 NatML.
*/

namespace NatSuite.ML.Vision {

    using System;
    using System.Threading.Tasks;
    using NatSuite.ML;
    using NatSuite.ML.Features;
    using NatSuite.ML.Internal;

    /// <summary>
    /// Image classifier for our super-important top-secret new product.
    /// </summary>
    public sealed class ResNet18HubPredictor : IMLAsyncPredictor<(string label, float score)> {

        #region --Client API--
        /// <summary>
        /// Create the ResNet18 predictor.
        /// </summary>
        /// <param name="model">ML model.</param>
        public ResNet18HubPredictor (MLModel model) => this.model = model as MLHubModel;

        /// <summary>
        /// Classify an image.
        /// </summary>
        /// <param name="inputs">Input image.</param>
        /// <returns>Predicted label with confidence score.</returns>
        public async Task<(string label, float score)> Predict (params MLFeature[] inputs) {
            // Check inputs
            if (inputs.Length != 1 || !(inputs[0] is MLImageFeature))
                throw new ArgumentException("ResNet18HubPredictor expects a single image feature", nameof(inputs));
            // Predict
            var hubFeature = (inputs[0] as IMLHubFeature).Serialize();
            var results = await model.Predict(hubFeature);
            // Marshal
            var label = (results[0] as MLTextFeature).text;
            var score = (results[1] as MLArrayFeature<float>)[0];
            return (label, score);
        }
        #endregion


        #region --Operations--
        private readonly MLHubModel model;

        void IDisposable.Dispose () { }
        #endregion
    }
}

File: Node/src/ResNet18HubPredictor.ts
/**
 *   resnet18
 *   Copyright (c) 2021 NatML.
 */
import {
  MLModel,
  MLHubModel,
  IMLAsyncPredictor,
  MLFeature,
  MLArrayFeature,
  MLImageFeature,
  MLTextFeature,
} from "natml";

/**
 * Image classifier for our super-important top-secret new product.
 * Given an image, the predictor returns the class label and confidence score.
 *
 * NB: this must be the default export, and it must not be renamed, for Webpack
 * to bundle it properly for use in a <script> tag in the browser through unpkg.
 */
export default class ResNet18HubPredictor
  implements IMLAsyncPredictor<[string, number]>
{
  private readonly model: MLHubModel;

  /**
   * Create the predictor.
   * @param model ML model.
   */
  public constructor(model: MLModel) {
    this.model = model as MLHubModel;
  }

  /**
   * Make a prediction.
   * @param features Input features.
   * @returns Prediction output.
   */
  public async predict(...features: MLFeature[]) {
    // Check inputs
    if (features.length !== 1 || !(features[0] instanceof MLImageFeature))
      throw new Error(`ResNet18HubPredictor expects a single image feature`);

    // Predict
    const hubFeature = await features[0].serialize();
    const results = await this.model.predict(hubFeature);

    // Marshal
    const label = await new MLTextFeature(results[0]).text;
    const score = (await new MLArrayFeature(results[1]).data)[0];

    return [label, score] as [string, number];
  }
}
